src/Microsoft.OData.Client/Serialization/HttpClientRequestMessage.cs

```diff
@@ -71,25 +72,43 @@ public class HttpClientRequestMessage : DataServiceClientRequestMessage, ISendin
         /// <summary>
         /// Constructor for HttpClientRequestMessage.
         /// Initializes the <see cref="HttpClientRequestMessage"/> class.
-        /// The args.ActualMethod is the actual method. In post tunneling situations method will be POST instead of the specified verb method.
-        /// The args.method is the specified verb method
         /// </summary>
-        /// </summary>
-        public HttpClientRequestMessage(DataServiceClientRequestMessageArgs args) 
+        public HttpClientRequestMessage(DataServiceClientRequestMessageArgs args)
             : base(args.ActualMethod)
         {
             _messageStream = new MemoryStream();
-            _handler = new HttpClientHandler();
-            _client = new HttpClient(_handler, disposeHandler: true);
-            _contentHeaderValueCache = new Dictionary<string, string>();
-            _effectiveHttpMethod = args.Method;
-            _requestUrl = args.RequestUri;
-            _requestMessage = new HttpRequestMessage(new HttpMethod(this.ActualMethod), _requestUrl);
+            _client = args.HttpClient;
+            try
+            {
+                if (_client == null)
+                {
+                    _handler = new HttpClientHandler();
+                    _client = new HttpClient(_handler, disposeHandler: true);
+                }
+                else
+                {
+#if (NETCOREAPP || NETSTANDARD)
+                    _handler = (HttpClientHandler)(_client.GetType().BaseType.GetField("_handler", BindingFlags.NonPublic | BindingFlags.Instance).GetValue(_client) as HttpMessageHandler);
+#else
+                    _handler = _client.GetType().BaseType.GetField("handler", BindingFlags.NonPublic | BindingFlags.Instance).GetValue(_client) as HttpClientHandler;

```

Comment:
How come it's `.GetField("handler", ...)` on this line while 2 lines above it's `.GetField("_handler", ...)`?

Reply:
This depends on the HttpClient implementation on .NETCORE AND .NET. In .netcore the field has the underscore while in .net it doesn't.

Reply:
Is there another alternative that does not tie us to private implementation details like this? This seems very brittle.

Reply:
If we really need the `_handler` instance, maybe we should inject `HttpClientHandler` instead of `HttpClient` and then use the injected `HttpClientHandler` to create a new `HttpClient` each time. We should avoid relying on private fields.

Reply:
Sinc we're using the `_handler` to make some configuration possible like the `Credentials` property, maybe we should just throw an UnsupportedException or InvalidOperationException if the developer tries to access that property when they have injected a client. The message could mention that in such a scenario, those settings should be done on the handler directly before injecting the `HttpClient` to the DataServicContext.